consensus/hotstuff/timeoutcollector/timeout_processor.go

```diff
@@ -0,0 +1,300 @@
+		notifier:  notifier,
+		partialTCTracker: accumulatedWeightTracker{
+			minRequiredWeight: timeoutThreshold,
+			done:              *atomic.NewBool(false),
+		},
+		tcTracker: accumulatedWeightTracker{
+			minRequiredWeight: qcThreshold,
+			done:              *atomic.NewBool(false),
+		},
+		sigAggregator:   sigAggregator,
+		newestQCTracker: NewNewestQCTracker(),
+	}, nil
+}
+
+// Process performs processing of timeout object in concurrent safe way. This
+// function is implemented to be called by multiple goroutines at the same time.
+// Design of this function is event driven, as soon as we collect enough weight
+// to create a TC or a partial TC we will immediately do so and submit it
+// via callback for further processing.
+// Expected error returns during normal operations:
+// * ErrTimeoutForIncompatibleView - submitted timeout for incompatible view
+// * model.InvalidTimeoutError - submitted invalid timeout(invalid structure or invalid signature)
+// * model.ErrViewForUnknownEpoch if no epoch containing the given view is known
+// All other errors should be treated as exceptions.
+func (p *TimeoutProcessor) Process(timeout *model.TimeoutObject) error {
+	if p.view != timeout.View {
+		return fmt.Errorf("received incompatible timeout, expected %d got %d: %w", p.view, timeout.View, ErrTimeoutForIncompatibleView)
+	}
+
+	if p.tcTracker.Done() {
+		return nil
+	}
+
+	err := p.validateTimeout(timeout)
+	if err != nil {
+		return fmt.Errorf("received invalid timeout: %w", err)
+	}
+
+	p.newestQCTracker.Track(timeout.NewestQC)
+
+	totalWeight, err := p.sigAggregator.VerifyAndAdd(timeout.SignerID, timeout.SigData, timeout.NewestQC.View)
+	if err != nil {
+		return fmt.Errorf("could not process invalid signature: %w", err)
+	}
+
+	if p.partialTCTracker.Track(totalWeight) {
+		p.notifier.OnPartialTcCreated(p.view, p.newestQCTracker.NewestQC(), timeout.LastViewTC)
+	}
+
+	// Checking of conditions for building TC are satisfied when willBuildTC is true.
+	// At this point, we have enough signatures to build a TC. Another routine
+	// might just be at this point. To avoid duplicate work, Track returns true only once.
+	willBuildTC := p.tcTracker.Track(totalWeight)
+
+	if !willBuildTC {
+		// either we do not have enough timeouts to build a TC, or another thread
+		// has already passed this gate and created a TC
+		return nil
+	}
+
+	tc, err := p.buildTC()
+	if err != nil {
+		return fmt.Errorf("internal error constructing TC: %w", err)
+	}
+	p.notifier.OnTcConstructedFromTimeouts(tc)
+
+	return nil
+}
+
+// validateTimeout performs validation of timeout object, verifies if timeout is correctly structured
+// and included QC and TC is correctly structured and signed.
+// ATTENTION: this function doesn't check if timeout signature is valid, this check happens in signature aggregator
+// Expected error returns during normal operations:
+// * model.InvalidTimeoutError - submitted invalid timeout(invalid structure or invalid signature)
```

Comment:
> invalid structure or **invalid signature**
we not verify the signature here. Following my earlier suggestion to neither check the signer ID nor the signature here:

```suggestion
// ATTENTION: this function does _not_ check whether the TO's `SignerID` is an authorized node nor if 
// the signature is valid. These checks happens in signature aggregator.
// Expected error returns during normal operations:
// * model.InvalidTimeoutError - submitted invalid timeout
```